packages/fast-components/src/design-system-provider/design-system-provider.ts

```diff
@@ -1,98 +1,271 @@
+            });
+        });
+    };
 
-    @designSystemProperty
-    public cornerRadius: number = 2;
+    /**
+     * Syncronizes the provider's design system with the local
+     * overrides. Any value defined on the instance will take priority
+     * over the value defined by the provider
+     */
+    private syncDesignSystemWithProvider(): void {
+        if (this.provider) {
+            Object.keys(this.provider.designSystem).forEach(key => {
+                if (this[key] === void 0) {
+                    this.designSystem[key] = this.provider!.designSystem[key];
+                }
+            });
+        }
+    }
 
-    @designSystemProperty
-    public disabledOpacity = 0.3;
+    /**
+     * Invoked when the provider observable property defined by the consumer is changed
+     * @param prev the previous value
+     * @param next the next value
+     */
+    private providerChanged(
+        prev: DesignSystemProvider | null,
+        next: DesignSystemProvider | null
+    ): void {
+        if (prev instanceof HTMLElement) {
+            Observable.getNotifier(prev).unsubscribe(this, designSystemKey);
+        }
 
-    @designSystemProperty
-    public density = 0;
+        if (next instanceof HTMLElement && isDesignSystemProvider(next)) {
+            Observable.getNotifier(next).subscribe(this, designSystemKey);
+            this.syncDesignSystemWithProvider();
+        }
+    }
+}
 
-    @designSystemProperty
-    public designUnit = 4;
+/**
+ * Simple proxy object to notifiy observers on
+ * object mutation.
+ * @param source The object to fire notifications
+ * @param key The observable property name to fire a notification on
+ */
+function mutationObserver(source: any, key: string): {} {
+    const notifier = Observable.getNotifier(source);
 
-    @designSystemProperty
-    public elevatedCornerRadius = 4;
+    return new Proxy(
+        {},
+        {
+            set(obj: any, prop: string, value: any): boolean {
+                obj[prop] = value;
 
-    @designSystemProperty
-    public focusOutlineWidth = 2;
+                notifier.notify(source, key);
+                return true;
+            },
+        }
+    );
+}
+
+function setCustomPropertyFactory(
+    source: any
+): (definition: CustomPropertyDefinition) => void {
+    let store: CustomPropertyDefinition[] = [];
+    let ticking = false;
+
+    return (definition: CustomPropertyDefinition) => {
+        const index = store.findIndex(x => x.name === definition.name);
+
+        if (index !== -1) {
+            store[index] = definition;
+        } else {
+            store.push(definition);
+        }
 
-    @designSystemProperty
-    public outlineWidth = 1;
+        if (ticking) {
+            return;
+        } else {
+            ticking = true;
 
-    @observable
-    public designSystem: DesignSystem = DesignSystemDefaults;
+            window.requestAnimationFrame(() => {
+                ticking = false;

```

Comment:
@EisenbergEffect This seemed like an opportunity to hook into the rendering engine but I'm not sure how to do that. My thought here was that setting individual property values was more efficient than concating a giant list of properties into a style attribute string and binding to that in the templating engine but I could be off-base on that.

Reply:
There's a method `DOM.queueUpdate` that you can use to hook into the render batch. That will let you remove the rAF code. You will still probably want to use your `ticking` guard to dedupe queuing things up.